Sources/NIOSOCKS/Messages/ClientGreeting.swift

```diff
@@ -0,0 +1,71 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// The SOCKS handshake begins with the client sending a greeting
+/// containing supported authentication methods.
+struct ClientGreeting: Hashable {
+    
+    /// The SOCKS protocol version - we currently only support v5.
+    public let version: UInt8 = 5
+    
+    /// The client's supported authentication methods, defined in RFC 1928.
+    public var methods: [AuthenticationMethod]
+    
+    /// Creates a new client greeting with the given authentication methods.
+    /// - parameter methods: The client's supported authentication methods.
+    public init(methods: [AuthenticationMethod]) {
+        self.methods = methods
+    }
+    
+    init?(buffer: inout ByteBuffer) throws {
+        guard
+            let version = buffer.readInteger(as: UInt8.self),
+            let numMethods = buffer.readInteger(as: UInt8.self)
+        else {
+            return nil
```

Comment:
this is missing a reset of `buffer = save` if you read short. other places too.

Reply:
We reset at the top level - this is an extremely simple parser, if one thing fails the entire thing does. It's neater to do it at the top level rather than restore at multiple points with the same end goal.

Reply:
usually functions have two modes:
- they fail (and leave arguments in alone)
- they success (do what it says)

your implementation violates that principle and introduces a half-failure mode where you modify the arguments. I'd definitely call this an anti-pattern and it's definitely not how any of the other stuff works.

Reply:
You make an excellent point, I'l rework it